### DISCUSSION TITLE
Having trouble remapping `<leader>c` to `gcc`

### DISCUSSION THREAD

**Discussion by refaelsh** [nix-community/nixvim#2089] [Q&A] [answered]
I have the following in my config:
```nix
programs.nixvim = {
  globals = {
    mapleader = ",";
    maplocalleader = ",";
  };
  keymaps = [
    {
      mode = "n";
      key = "<leader>c";
      action = "gcc";
      options.noremap = true;
      options.silent = true;
    }
    {
      mode = "v";
      key = "<leader>c";
      action = "gc";
      options.noremap = true;
      options.silent = true;
    }
  ];
};
```

Pressing `gcc` obviously works, but pressing `<leader>c` does not work.
Please help :-)

Thanks in advance.

**Comment by MattSturgeon** [comment]
I think `noremap` may be the issue here, since presumably `gcc` comes from a plugin and therefore _is_ a remap.

**Reply by refaelsh** [reply]
No, `gcc` does not come from a plugin, it is now part of Neovim, built-in. Here is the link that says so: https://github.com/neovim/neovim/pull/28176.
Also, I tried with and without `noremap`.

**Reply by MattSturgeon** [reply]
> No, `gcc` does not come from a plugin, it is now part of Neovim, built-in. Here is the link that says so: [neovim/neovim#28176](https://github.com/neovim/neovim/pull/28176).

TIL; thanks for educating me :grin: 

---

In that case this doesn't make much sense to me. Out of curiosity, does which-key's "proxy" mapping handle it any differently?

Beyond that, if none of us can help, you may need to seek help somewhere more nvim specific.

**Reply by refaelsh** [reply]
> does which-key's "proxy" mapping handle it any differently?

I have no idea what you just said :-)

**Reply by MattSturgeon** [reply]
Which-key has a `proxy` [mapping option](https://github.com/folke/which-key.nvim/?tab=readme-ov-file#%EF%B8%8F-mappings), used to create aliases of other mappings.

**Reply by refaelsh** [reply]
> Which-key has a `proxy` [mapping option](https://github.com/folke/which-key.nvim/?tab=readme-ov-file#%EF%B8%8F-mappings), used to create aliases of other mappings.

No idea. I maybe heard about `which-key` in the distant past, but that's it, that's the extent of my `which-key` knowledge.

**Comment by refaelsh** [comment]
I took a look at the generated lua file, and heere is what I see:
```lua
do
    local __nixvim_binds = {
        { action = ";", key = ":", mode = "n" },
        { action = ":", key = ";", mode = "n" },
        { action = "P", key = "p", mode = "v" },
        { action = ":noh<cr>", key = "<leader><space>", mode = "n" },
        { action = "<nop>", key = "<left>", mode = "n" },
        { action = "<nop>", key = "<right>", mode = "n" },
        { action = "<nop>", key = "<up>", mode = "n" },
        { action = "<nop>", key = "<down>", mode = "n" },
        { action = "<nop>", key = "<left>", mode = "i" },
        { action = "<nop>", key = "<right>", mode = "i" },
        { action = "<nop>", key = "<up>", mode = "i" },
        { action = "<nop>", key = "<down>", mode = "i" },
        { action = "<nop>", key = "<left>", mode = "v" },
        { action = "<nop>", key = "<right>", mode = "v" },
        { action = "<nop>", key = "<up>", mode = "v" },
        { action = "<nop>", key = "<down>", mode = "v" },
        { action = "gcc", key = "<leader>c", mode = "n" },
        { action = "gc", key = "<leader>c", mode = "v" },
        { action = "gj", key = "j", mode = "n" },
        { action = "gk", key = "k", mode = "n" },
        { action = "o<ESC>", key = "<Enter>", mode = "n" },
        { action = "O<ESC>", key = "<S-Enter>", mode = "n" },
        { action = "i#[test]<CR>fn () {<CR>}<ESC>kwi", key = ",t", mode = "n" },
        {
            action = "i#[cfg(test)]<CR>mod tests {<CR>use super::*;<CR><CR>#[test]<CR>fn () {<CR>}<CR><ESC>xxxxi}<ESC>kkwwi",
            key = ",tm",
            mode = "n",
        },
        { action = "<C-d>zz", key = "<C-d>", mode = "n" },
        { action = "<C-u>zz", key = "<C-u>", mode = "n" },
        { action = "nzzzv", key = "n", mode = "n" },
        { action = "Nzzzv", key = "N", mode = "n" },
    }
    for i, map in ipairs(__nixvim_binds) do
        vim.keymap.set(map.mode, map.key, map.action, map.options)
    end
end
```
I am not that expereinced with Lua, but everything looks good.
Also, every key mapping from the above code works, except `<leader>c`.
Very interesting!
In any case, I will continue googling, and if I find the solution, I will post it here.

**Reply by MattSturgeon** [reply]
Do you have the same issue manually creating the mapping (e.g. using `:map`). It's starting to feel like a bug in nvim, or at least a quirk I'm unaware of.

**Comment by refaelsh** [comment]
> Do you have the same issue manually creating the mapping

I just checked, this way its working.

I am starting to think that some plugin re-maps `<leader>c`.
Also, when running this: `:verbose nmap <buffer> <leader>c`, I get this: `No mapping found`.

I will start disabling plugins one by one.

**Reply by refaelsh** [reply]
I got my nixvim config to the bare minimum:
```nix
programs.nixvim = {
  enable = true;
  globals = {
    mapleader = " ";
    maplocalleader = " ";
  };
  keymaps = [
    {
      mode = "n";
      key = "<leader>c";
      action = "gcc";
    }
  ];
};
```
`gcc` works and `<leader>c` does not.
I am willing to try anything :-)

**Reply by traxys** [reply]
I'd suggest extracting the lua code of the configuration, checking that `nvim -u <the config>` has the same issue, and asking around a pure neovim community, as this seems mostly unrelated to nixvim

**Reply by refaelsh** [reply]
> I'd suggest extracting the lua code of the configuration, checking that `nvim -u <the config>` has the same issue, and asking around a pure neovim community, as this seems mostly unrelated to nixvim

Great suggestion. Thank you. I will do it.

**Reply by refaelsh** [reply]
For future googlers: here is the stackoverflow question I opened: https://stackoverflow.com/questions/78926822/cant-map-leaderc-in-a-minimal-init-lua.

**Comment by refaelsh** [comment]
I discovered something new. I don't know if it is relevant, but I will post it here anyway, maybe it will light a light bulb.
I have the following in my config:
```nix
lsp = {
  enable = true;
  keymaps.lspBuf = {
    "<leader>d" = "definition";
    "<leader>a" = "code_action";
    "<leader>i" = "implementation";
    "<leader>ic" = "incoming_calls";
    "<leader>f" = "format";
    "<leader>h" = "hover";
    "<leader>r" = "rename";
  };
};
```
Out of the above mappings only `<leader>d` works. No idea why.

**Reply by khaneliman** [reply]
> I discovered something new. I don't know if it is relevant, but I will post it here anyway, maybe it will light a light bulb. I have the following in my config:
> 
> ```nix
> lsp = {
>   enable = true;
>   keymaps.lspBuf = {
>     "<leader>d" = "definition";
>     "<leader>a" = "code_action";
>     "<leader>i" = "implementation";
>     "<leader>ic" = "incoming_calls";
>     "<leader>f" = "format";
>     "<leader>h" = "hover";
>     "<leader>r" = "rename";
>   };
> };
> ```
> 
> Out of the above mappings only `<leader>d` works. No idea why.

All the different lsp method calls require a supporting lsp server that has the registered capabilities. Not every lsp server supports every capability and they are context specific actions for most of them so it might not support it on the specific code you have cursor on.

**Reply by refaelsh** [reply]
> All the different lsp method calls require a supporting lsp server that has the registered capabilities. Not every lsp server supports every capability and they are context specific actions for most of them so it might not support it on the specific code you have cursor on.

Good catch! That was it. I tried a language server that had its formatting misconfigured.

Thank you :-)

**Comment by refaelsh** [comment, marked as answer]
Issue solved. Like this:
```nix
{
  mode = "n";
  key = "<leader>c";
  action = "gcc";
  options.remap = true;
}
```
I found the answer here: https://www.reddit.com/r/neovim/comments/1d278fz/keymap_comment_nvim_010/.
In case the link stops working in the distant future, here is a quote:
> If you are wondering why remap is necessary, it is because gcc is a mapping as well. So when remap is false, it will not recursively map to other mappings. You can see it and other defaults here https://github.com/neovim/neovim/blob/master/runtime/lua/vim/_defaults.lua#L135
